### PR TITLE
Allow skipping a prefix of the eras in Cardano mode

### PR DESCRIPTION
**Pull request by mrBliss**
Add support for immediately executing one or more hard forks during initialisation, when configured using `TriggerHardForkAtEpoch 0`. This will allow starting a node in Cardano mode directly in the Shelley, Allegra, or Mary era without having to spend an epoch in each preceding era.

For example, to start a Cardano node in the Mary era, add the following to your `cardano-node` configuration:

```
TestShelleyHardForkAtEpoch: 0
TestAllegraHardForkAtEpoch: 0
TestMaryHardForkAtEpoch:    0
```

Registering initial staking and funds is also allowed in Cardano mode, but will only take effect if configured to skip at least the Byron era.

### CODE REVIEW

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 1]
```diff
@@ -564,7 +598,48 @@ protocolInfoCardano protocolParamsByron@ProtocolParamsByron {
...
+         ) shelleyEra
+      -> (     State.Current LedgerState
+          -.-> State.Current LedgerState
+         ) (ShelleyBlock shelleyEra)
+    unwrapped g = fn $ \(State.Current start state) ->
+          State.Current start
+        . unComp
+        . apFn g
+        . Comp
+        $ state
```
Comment (mrBliss) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:639]: Do the lifting of `State.Current` separately using https://github.com/input-output-hk/ouroboros-network/blob/376a5ba3a4e5a9c1dc68bdc1c64d875f3762abfe/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/State/Lift.hs#L18

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 2]
```diff
@@ -476,3 +412,85 @@ instance ShelleyBasedEra era => NodeInitStorage (ShelleyBlock era) where
...
+
+-- | Register the initial staking and initial funds in the 'SL.NewEpochState'.
+--
+-- HERE BE DRAGONS! This function is intended to help in testing.
+--
+-- In production, the genesis should /not/ contain any initial staking or
+-- initial funds.
+--
+-- TODO adapt and reuse @registerGenesisStaking@ from @cardano-ledger-specs@.
+registerGenesisStaking ::
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:428]: Discuss this with @nfrisby (unioning).

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 3]
```diff
@@ -564,7 +598,48 @@ protocolInfoCardano protocolParamsByron@ProtocolParamsByron {
...
+      -> LedgerState (ShelleyBlock era)
+     )
+  -> LedgerState (CardanoBlock c)
+  -> LedgerState (CardanoBlock c)
+overShelleyBasedLedgerState f (HardForkLedgerState (State.HardForkState st)) =
+    -- We can't use the pattern synonyms to match and /construct/ the various
+    -- 'LedgerState's because they are not bidirectional. Moreover, such a
+    -- solution would need to be extended for /each/ era. Instead, we do it
+    -- generically for all Shelley-based eras and then cons Byron in front.
+    HardForkLedgerState $ State.HardForkState $ hap fs st
```
Comment (mrBliss) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:618]: `hap` on the `State.HardForkState`

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 4]
```diff
@@ -484,3 +488,57 @@ registerGenesisStaking staking nes = nes {
...
           pState'
+
+-- | Register the initial funds in the 'SL.NewEpochState'.
+--
+-- HERE BE DRAGONS! This function is intended to help in testing.
+--
+-- In production, the genesis should /not/ contain any initial funds.
+--
+-- The given funds are /added/ to the existing UTxO.
+--
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:499]: Property:
```
   genesisUTxO genesis
== <genesisUTxO'> (sgInitialFunds genesis) -- for some <genesisUTxO'>
== <extractUTxO> (registerInitialFunds (sgInitialFunds genesis) <empty NewEpochState>)
```

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 5]
```diff
@@ -484,3 +488,57 @@ registerGenesisStaking staking nes = nes {
...
+    accountState = SL.esAccountState epochState
+    ledgerState  = SL.esLState       epochState
+    utxoState    = SL._utxoState     ledgerState
+    utxo         = SL._utxo          utxoState
+    reserves     = SL._reserves      accountState
+
+    initialFundsMapping :: Map (SL.TxIn (EraCrypto era)) (SL.TxOut era)
+    initialFundsMapping = Map.fromList [
+          (txIn, txOut)
+        | (addr, amount) <- Map.toList initialFunds,
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:523]: Comma!

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 6]
```diff
@@ -484,3 +488,57 @@ registerGenesisStaking staking nes = nes {
...
+           let txIn  = SL.initialFundsPseudoTxIn addr
+               txOut = SL.TxOut addr (inject amount)
+        ]
+
+    -- 'Map.union' is left-biased, so 'initialFundsMapping' has to be the
+    -- \"left\" argument.
+    utxo' = SL.UTxO $ Map.union initialFundsMapping (SL.unUTxO utxo)
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:530]: `error` when there's overlap

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 7]
```diff
@@ -484,3 +488,57 @@ registerGenesisStaking staking nes = nes {
...
           dState'
           pState'
+
+-- | Register the initial funds in the 'SL.NewEpochState'.
+--
+-- HERE BE DRAGONS! This function is intended to help in testing.
+--
+-- In production, the genesis should /not/ contain any initial funds.
+--
+-- The given funds are /added/ to the existing UTxO.
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:498]:
```
PRECONDITION: the given funds must not be part of the existing UTxO.
> forall (addr, _) in initialFunds.
>    Map.notElem (SL.initialFundsPseudoTxIn addr) existingUTxO
```

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 8]
```diff
@@ -484,3 +488,57 @@ registerGenesisStaking staking nes = nes {
...
+    -- 'Map.union' is left-biased, so 'initialFundsMapping' has to be the
+    -- \"left\" argument.
+    utxo' = SL.UTxO $ Map.union initialFundsMapping (SL.unUTxO utxo)
+
+    -- How much extra should be subtracted from the reserves?
+    extraOutOfReserves = coin (SL.balance utxo' <-> SL.balance utxo)
+
+    -- Update the reserves
+    accountState' = accountState {
+          SL._reserves = reserves <-> extraOutOfReserves
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:537]: Without the overlap, we can just subtract the balance of `utxo'`.

**Review by edsko — APPROVED** [review]

**Review by deepfire — APPROVED** [review]

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**